# Линейное программирование

Линейное программирование — это метод оптимизации для системы линейных ограничений и линейной целевой функции. Целевая функция определяет оптимизируемую величину, и цель линейного программирования состоит в том, чтобы найти значения переменных, которые максимизируют или минимизируют целевую функцию.

Линейное программирование полезно применять для многих задач, требующих оптимизации ресурсов:

- В производстве — чтобы рассчитать человеческие и технические ресурсы и минимизировать стоимость итоговой продукции.

- При составлении бизнес-планов — чтобы решить, какие продукты продавать и в каком количестве, чтобы максимизировать прибыль.

- В логистике — чтобы определить, как использовать транспортные ресурсы для выполнения заказов за минимальное время.

- В сфере общепита — чтобы составить расписание для официантов.

Задача линейного программирования — это задача оптимизации, в которой целевая функция и функции-ограничения линейны, а все переменные неотрицательны.

В языке Python есть множество библиотек, с помощью которых можно решить задачу линейного программирования. Вот основные, которые мы рассмотрим в данном юните:

- SciPy (scipy.optimize.linprog);
- CVXPY; 
- PuLP.

### Обычно в коде существуют такие обозначения:

--- 
1. c - коэффициенты целевой функции.

Пример: $P(x, y) = c_1x + c_2y$

Если мы **максимизируем** прибыль, нам нужно записать коэффициенты с отрицательным знаком, так как linprog решает только задачи минимизации.

Например, если мы хотим максимизировать $4x + 3y$, то мы передаем:

``` python
c = [-4, -3]
```
---

2. A - коэффициенты ограничений.

где:

- A — матрица коэффициентов перед переменными x, y

- b — правая часть ограничений.

Пример: есть два ограничения:

- $2x + 4y <= 220$
- $3x + 2y <= 150$

```python
A = [
    [2, 4],  # 2x + 4y <= 220
    [3, 2]   # 3x + 2y <= 150
]

b = [220, 150]
```
---
3. x_bounds и y_bounds — границы переменных.

Они задают диапазон значений для x и y

Пример: x >= 0 и y >= 0

```python
x_bounds = (0, None)  # x не может быть отрицательным, но верхней границы нет
y_bounds = (0, None)  # y не может быть отрицательным
```

Если у x есть верхнее ограничение (например, 0 <= x <= 50), то:

```python
x_bounds = (0, 50)
```

---

### Пример c помощью: (SciPy (scipy.optimize.linprog))

У нас есть 6 товаров с заданными ценами на них и заданной массой.

Вместимость сумки, в которую мы можем положить товары, заранее известна и равна 15 кг.

Какой товар и в каком объёме необходимо взять, чтобы сумма всех цен товаров была максимальной?

In [1]:
from scipy.optimize import linprog
import numpy as np

values = [4, 2, 1, 7, 3, 6] #стоимости товаров
weights = [5, 9, 8, 2, 6, 5] #вес товаров
C = 15 # вместимость сумки
n = 6 # кол-во товаров

c = - np.array(values) # изменяем знак, чтобы перейти от задачи максимизации к задаче минимизации
A = np.array(weights) # конвертируем список с весами в массив

# Здесь нам необходимо вспомнить линейную алгебру, 
# так как очень важно, чтобы векторы были в нужных нам размерностях, 
# иначе мы не сможем использовать матричное умножение. 
# Вектор A размера 6 мы превращаем в матрицу размера (1, 6) с помощью функции expand_dims(). 
# Создаём все необходимые переменные:

A = np.expand_dims(A, 0) # преобразуем размерность массива
b = np.array([C]) # конвертируем вместимость в массив

result = linprog(c=c, A_ub=A, b_ub=b)

# Выводим результат
print(f'{result}\n')
print("Максимальная сумма всех цен товаров:", -result.fun)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: -52.5
              x: [ 0.000e+00  0.000e+00  0.000e+00  7.500e+00  0.000e+00
                   0.000e+00]
            nit: 0
          lower:  residual: [ 0.000e+00  0.000e+00  0.000e+00  7.500e+00
                              0.000e+00  0.000e+00]
                 marginals: [ 1.350e+01  2.950e+01  2.700e+01  0.000e+00
                              1.800e+01  1.150e+01]
          upper:  residual: [       inf        inf        inf        inf
                                    inf        inf]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              0.000e+00  0.000e+00]
          eqlin:  residual: []
                 marginals: []
        ineqlin:  residual: [ 0.000e+00]
                 marginals: [-3.500e+00]
 mip_node_count: 0
 mip_dual_bound: 0.0
        mip_gap: 0.0

Максимальная сумма всех цен

Получаем искомое значение функции — 52.5 (в выводе значение с минусом, но мы меняем знак, возвращаясь к задаче максимизации). $x = (0, 0 ,0, 7.5, 0, 0)$. Таким образом, мы взяли только самую дорогую, четвёртую вещь. Она одна весит 2 кг, а если взять её 7.5 раз, то получится как раз 15 кг. Отлично, задача решена.

### Пример: 

Магазин спортивных товаров получает прибыль в размере 6 долларов с каждого проданного футбольного мяча и 5.5 долларов — с бейсбольного.

Каждый месяц магазин продаёт от 35 до 45 футбольных мячей и от 40 до 55 бейсбольных.

Известно, что в этом месяце у магазина есть в общей сложности 80 мячей.

Какую максимальную прибыль в этом месяце может получить магазин от продажи мячей?

In [2]:
# Коэффициенты целевой функции (мы максимизируем, поэтому берем отрицательные значения)
c = [-6, -5.5]  # (футбольные мячи, бейсбольные мячи)

# Ограничения
A = [[1, 1],    # x + y = 80 (в этом месяце)
     [1, 0],    # x <= 45
     [-1, 0],   # x >= 35
     [0, 1],    # y <= 55
     [0, -1]]   # y >= 40

b = [80, 45, -35, 55, -40] # Правая часть неравенств

# Границы переменных
x_bounds = (35, 45)
y_bounds = (40, 55)

# Решаем задачу
result = linprog(c, A_ub=A, b_ub=b, bounds=[x_bounds, y_bounds], method="highs")

# Выводим результат
print("Оптимальное количество футбольных мячей:", result.x[0])
print("Оптимальное количество бейсбольных мячей:", result.x[1])
print("Максимальная прибыль:", -result.fun)

Оптимальное количество футбольных мячей: 40.0
Оптимальное количество бейсбольных мячей: 40.0
Максимальная прибыль: 460.0


---

### Пример c помощью: (CVXPY)

Снова решим задачу из примера № 1, но уже предположим, что товары нельзя дробить, и будем решать задачу целочисленного линейного программирования.

SciPy не умеет решать такие задачи, поэтому будем использовать новую библиотеку CVXPY.

In [3]:
import cvxpy

values = [4, 2, 1, 7, 3, 6] #стоимости товаров
weights = [5, 9, 8, 2, 6, 5] #вес товаров
C = 15 # вместимость сумки
n = 6 # кол-во товаров

c = - np.array(values) # изменяем знак, чтобы перейти от задачи максимизации к задаче минимизации
A = np.array(weights) # конвертируем список с весами в массив
A = np.expand_dims(A, 0) # преобразуем размерность массива
b = np.array([C]) # конвертируем вместимость в массив

Поскольку мы решаем теперь задачу целочисленными значениями, нам необходимо это явно указать - x целочисленные значения.

In [4]:
x = cvxpy.Variable(shape=n, integer=True) # размерность n = 6, целочисленные значения = True

In [5]:
constrains = (A @ x <= b) # Устанавливаем ограничение на вес товаров, вес не должен превышать 15
x_positive = (x >= 0) # Также указываем, что x может быть только положительным или равен 0
total_value = c @ x # переменная для вычисления веса где вес * на количество товаров

# Оптимизационная задача (минимизируем значения и указываем содержание ограничения)
problem = cvxpy.Problem(cvxpy.Minimize(total_value), constraints=[constrains, x_positive])

# Решаем задачу
result = problem.solve()
print(f'Максимальная сумма товаров составляет: {- result}')

Максимальная сумма товаров составляет: 49.0


In [6]:
# при каких количествах товаров эта задача решена (берем 4-й товар 7 раз)
print(f'Количество товаров добавленных в сумку: {x.value}')

Количество товаров добавленных в сумку: [-0. -0. -0.  7. -0.  0.]


Здесь мы уже получаем 49, и берём только четвёртый товар в количестве семи штук. Можно увидеть, что результат, в целом, очень близок к первому, когда мы использовали библиотеку SciPy — различие лишь в добавлении целочисленности. Значит, у нас получилось решить задачу, когда мы добавили недостающее условие.

А что если мы можем брать не любое количество товаров, а только один или не брать их вовсе? Задаём x типа boolean.

$x = 1$ или $x = 0$ 

In [7]:
x = cvxpy.Variable(shape=n, boolean=True) # Здесь указываем True

constrains = (A @ x <= b) # Устанавливаем ограничение на вес товаров, вес не должен превышать 15
x_positive = (x >= 0) # Также указываем, что x может быть только положительным или равен 0
total_value = c @ x # переменная для вычисления веса где вес * на количество товаров

# Оптимизационная задача (минимизируем значения и указываем содержание ограничения)
problem = cvxpy.Problem(cvxpy.Minimize(total_value), constraints=[constrains, x_positive])

# Решаем задачу
result = problem.solve()
print(f'Максимальная сумма товаров составляет: {- result}')
# при каких количествах товаров эта задача решена (берем 4-й товар 7 раз)
print(f'Количество товаров добавленных в сумку: {x.value}') # Взяли 1-й, 4-й и 6-й товары по одному разу

Максимальная сумма товаров составляет: 17.0
Количество товаров добавленных в сумку: [1. 0. 0. 1. 0. 1.]


---

### Пример c помощью: (PuLP)

В нашей каршеринговой компании две модели автомобилей: модель A и модель B. Автомобиль A даёт прибыль в размере 20 тысяч в месяц, а автомобиль B — 45 тысяч в месяц. Мы хотим заказать на заводе новые автомобили и максимизировать прибыль. Однако на производство и ввод в эксплуатацию автомобилей понадобится время:

- Проектировщику требуется 4 дня, чтобы подготовить документы для производства каждого автомобиля типа A, и 5 дней — для каждого автомобиля типа B.

- Заводу требуется 3 дня, чтобы изготовить модель A, и 6 дней, чтобы изготовить модель B.

Менеджеру требуется 2 дня, чтобы ввести в эксплуатацию в компании автомобиль A, и 7 дней — B автомобиль B.

- Каждый специалист может работать суммарно 30 дней.

Заметьте, что здесь мы снова пишем обычные неравенства, а не условия в матричном виде. Дело в том, что для данной библиотеки так «удобнее», так как она принимает все условия в «первичном» виде.

In [ ]:
from pulp import *

# Создаем переменную нашей задачи, указывая, что мы хотим максимизировать функцию
problem = LpProblem('Производство автомобилей', LpMaximize)

# Записываем в более привычном виде две переменные (авто), где мы уже указваем то, 
# что они не могут быть отрицательными и значения целочисленные
A = LpVariable('Автомобиль А', lowBound=0, cat=LpBinary)
B = LpVariable('Автомобиль B', lowBound=0, cat=LpBinary)

# += просто добавляем новые условия к нашей переменной-задаче
# Целевая функция (максимизировать дохот от производства автомобилей)
problem += 20000*A + 45000*B

# Ограничения
problem += 4*A + 5*B <= 30
problem += 3*A + 6*B <= 30
problem += 2*A + 7*B <= 30

# Решение задачи
problem.solve()

# Вывод результата
print("Количество автомобилей модели А: ", A.varValue)
print("Количество автомобилей модели В: ", B.varValue)
print("Суммарный доход: ", value(problem.objective))

Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Users/alexander/.pyenv/versions/3.12.1/lib/python3.12/site-packages/pulp/solverdir/cbc/osx/64/cbc /var/folders/0g/dcvscgwn2bx5ldpwsnftw38c0000gn/T/af5b217179f349ccb38e65ffe2b2c4ee-pulp.mps -max -timeMode elapsed -branch -printingOptions all -solution /var/folders/0g/dcvscgwn2bx5ldpwsnftw38c0000gn/T/af5b217179f349ccb38e65ffe2b2c4ee-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 8 COLUMNS
At line 21 RHS
At line 25 BOUNDS
At line 28 ENDATA
Problem MODEL has 3 rows, 2 columns and 6 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Continuous objective value is 216667 - 0.00 seconds
Cgl0004I processed model has 3 rows, 2 columns (2 integer (0 of which binary)) and 6 elements
Cutoff increment increased from 1e-05 to 5000
Cbc0012I Integer solution of -195000 found by DiveCoefficient after 0 iterations and 0 nodes (0.00 seconds

/Users/alexander/.pyenv/versions/3.12.1/lib/python3.12/site-packages/pulp/pulp.py:1298: UserWarning: Spaces are not permitted in the name. Converted to '_'
  warnings.warn("Spaces are not permitted in the name. Converted to '_'")


Выходит, что необходимо произвести 1 автомобиль типа A и 4 автомобиля типа B. Тогда суммарный чистый доход будет равен 200 тысячам.

---

# Практика

<img src='Images/ml_13.png'>

In [17]:
# Создаем задачу линейного программирования
problem = LpProblem('Оптимальный план перевозок', LpMinimize)

# Постановка задачи Первый склад
x11 = LpVariable('Склад 1 - 1 ТЦ', lowBound=0, cat=LpInteger)
x12 = LpVariable('Склад 1 - 2 ТЦ', lowBound=0, cat=LpInteger)
x13 = LpVariable('Склад 1 - 3 ТЦ', lowBound=0, cat=LpInteger)

# Постановка задачи Второй склад
x21 = LpVariable('Склад 2 - 1 ТЦ', lowBound=0, cat=LpInteger)
x22 = LpVariable('Склад 2 - 2 ТЦ', lowBound=0, cat=LpInteger)
x23 =LpVariable('Склад 2 - 3 ТЦ', lowBound=0, cat=LpInteger)

# Целевая функция стоимость перевозки, которую нужно минимизировать
problem += 2*x11 + 5*x12 + 3*x13 + 7*x21 + 7*x22 + 6*x23, "Total cost"

# Ограничения запасам на складах
problem += x11 + x12 + x13 <= 180 # Склад 1
problem += x21 + x22 + x23 <= 220 # Склад 2

# Ограничения по потребностям торговых центров
problem += x11 + x21 == 110 # ТЦ 1
problem += x12 + x22 == 150 # ТЦ 2
problem += x13 + x23 == 140 # ТЦ 3

problem.solve()

# Вывод результата
print("Количество товаров со Склада 1 для ТЦ 1: ", x11.varValue)
print("Количество товаров со Склада 1 для ТЦ 2: ", x12.varValue)
print("Количество товаров со Склада 1 для ТЦ 3: ", x13.varValue)

print("Количество товаров со Склада 2 для ТЦ 1: ", x21.varValue)
print("Количество товаров со Склада 2 для ТЦ 2: ", x22.varValue)
print("Количество товаров со Склада 2 для ТЦ 3: ", x23.varValue)
print("Минимальная стоимость перевозки: ", round(value(problem.objective)))

Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Users/alexander/.pyenv/versions/3.12.1/lib/python3.12/site-packages/pulp/solverdir/cbc/osx/64/cbc /var/folders/0g/dcvscgwn2bx5ldpwsnftw38c0000gn/T/75ed9947f8414cb499c87a2dc5de24ec-pulp.mps -timeMode elapsed -branch -printingOptions all -solution /var/folders/0g/dcvscgwn2bx5ldpwsnftw38c0000gn/T/75ed9947f8414cb499c87a2dc5de24ec-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 10 COLUMNS
At line 41 RHS
At line 47 BOUNDS
At line 54 ENDATA
Problem MODEL has 5 rows, 6 columns and 12 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Continuous objective value is 1900 - 0.00 seconds
Cgl0004I processed model has 3 rows, 4 columns (4 integer (0 of which binary)) and 8 elements
Cutoff increment increased from 1e-05 to 0.9999
Cbc0012I Integer solution of 1900 found by DiveCoefficient after 0 iterations and 0 nodes (0.01 seconds)
Cbc0

/Users/alexander/.pyenv/versions/3.12.1/lib/python3.12/site-packages/pulp/pulp.py:1298: UserWarning: Spaces are not permitted in the name. Converted to '_'
  warnings.warn("Spaces are not permitted in the name. Converted to '_'")


<img src='Images/ml_14.png'>

In [22]:
# Создаем задачу линейного программирования
problem = LpProblem('О назначении исполнителей', LpMinimize)

# Определяем переменные (каждое значение - 1 или 0, назначена ли задача исполнителю)
x = [[LpVariable(f"x_{i}_{j}", cat=LpBinary) for j in range(5)] for i in range(5)]

# Стоимости выполнения задач исполнителями (из таблицы)
costs = [
    [1000, 12, 10, 19, 8],
    [12, 1000, 3, 7, 2],
    [10, 3, 1000, 6, 20],
    [19, 7, 6, 1000, 4],
    [8, 2, 20, 4, 1000]
]

# Целевая функция - минимизация суммарных затрат
problem += lpSum(costs[i][j] * x[i][j] for i in range(5) for j in range(5)), 'Total cost'

# Ограничение: Каждая задача должна быть назначена ровно одному исполнителю
for j in range(5):
    problem += lpSum(x[i][j] for i in range(5)) == 1

# Ограничение: Каждый исполнитель выполняет ровно одну задачу
for i in range(5):
    problem += lpSum(x[i][j] for j in range(5)) == 1
 
problem.solve()

# Вывод результата
for i in range(1, 6):
    for j in range(1, 6):
        if eval(f"x{i}{j}.varValue") == 1:
            print(f"Исполнитель {i} выполняет задачу {j}")

# Выводим минимальную стоимость работ
print("Минимальная стоимость работ:", round(value(problem.objective)))

Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Users/alexander/.pyenv/versions/3.12.1/lib/python3.12/site-packages/pulp/solverdir/cbc/osx/64/cbc /var/folders/0g/dcvscgwn2bx5ldpwsnftw38c0000gn/T/d918b0f6b6bb41d68977ed79178fd461-pulp.mps -timeMode elapsed -branch -printingOptions all -solution /var/folders/0g/dcvscgwn2bx5ldpwsnftw38c0000gn/T/d918b0f6b6bb41d68977ed79178fd461-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 15 COLUMNS
At line 141 RHS
At line 152 BOUNDS
At line 178 ENDATA
Problem MODEL has 10 rows, 25 columns and 50 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Continuous objective value is 32 - 0.00 seconds
Cgl0004I processed model has 10 rows, 25 columns (25 integer (25 of which binary)) and 50 elements
Cutoff increment increased from 1e-05 to 0.9999
Cbc0038I Initial state - 0 integers unsatisfied sum - 0
Cbc0038I Solution found of 32
Cbc0038I Before

<img src='Images/ml_15.png'>

In [24]:
# Создаем задачу линейного программирования (ЦЛП - целочисленного линейного программирования)
problem = LpProblem('Коммивояжер', LpMinimize)

# Определим точки
points = ['A', 'B', 'C', 'D', 'E']

# Определим количество точке
n = len(points)

# Расстояния между точками
distances = {
    ('A', 'B') : 12,
    ('A', 'C') : 10,
    ('A', 'D') : 19,
    ('A', 'E') : 8,
    ('B', 'C') : 3,
    ('B', 'D') : 7,
    ('B', 'E') : 2,
    ('C', 'D') : 6,
    ('C', 'E') : 20,
    ('D', 'E') : 4
}

# Создание переменных x_ij (0 или 1 - используем ли маршрут)
x = LpVariable.dicts('x', [(i, j) for i in points for j in points if i != j], cat=LpBinary)

# Вспомогательные переменные для исключения подциклов
u = LpVariable.dicts('u', points, lowBound=1, upBound=n, cat=LpInteger)

# Целевая функция - минимизируем длину маршрута
problem += lpSum(distances.get((i, j), distances.get((j, i))) * x[i, j] for i in points for j in points if i != j)

# Ограничение 1: Из каждой точки выходит ровно один маршрут
for i in points:
    problem += lpSum(x[i, j] for j in points if i != j) == 1
    
# Ограничение 2: В каждую точку входит ровно один маршрут

for j in points:
    problem += lpSum(x[i, j] for i in points if i != j) == 1
    
    # Ограничение 3: Исключение подциклов
for i in points:
    for j in points:
        if i != j and i != "A" and j != "A":  # A - стартовая точка
            problem += u[i] - u[j] + n * x[i, j] <= n - 1
            
# Решаем задачу
problem.solve()

# Вывод решения
print("Оптимальный маршрут:")
for i in points:
    for j in points:
        if i != j and x[i, j].varValue == 1:
            print(f"{i} -> {j}")

print("Минимальная длина маршрута:", value(problem.objective))

Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Users/alexander/.pyenv/versions/3.12.1/lib/python3.12/site-packages/pulp/solverdir/cbc/osx/64/cbc /var/folders/0g/dcvscgwn2bx5ldpwsnftw38c0000gn/T/07860ea293304da59a9a3b7330d231ba-pulp.mps -timeMode elapsed -branch -printingOptions all -solution /var/folders/0g/dcvscgwn2bx5ldpwsnftw38c0000gn/T/07860ea293304da59a9a3b7330d231ba-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 27 COLUMNS
At line 172 RHS
At line 195 BOUNDS
At line 224 ENDATA
Problem MODEL has 22 rows, 24 columns and 76 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Continuous objective value is 32 - 0.00 seconds
Cgl0003I 0 fixed, 0 tightened bounds, 12 strengthened rows, 0 substitutions
Cgl0003I 0 fixed, 0 tightened bounds, 12 strengthened rows, 0 substitutions
Cgl0004I processed model has 22 rows, 24 columns (24 integer (20 of which binary)) and 100 eleme